funsor/ops.py

```diff
@@ -303,6 +303,66 @@ def reciprocal(x):
 }
 
 
+# Linear algebra ops
+
+@Op
+def cholesky(x):
+    raise NotImplementedError
+
+
+@Op
+def cholesky_inverse(x):
+    raise NotImplementedError
+
+
+@Op
+def triangular_solve_op(x, y, upper, transpose):
```

Comment:
It is probably cleaner to encode parameterized ops as classes, as e.g. `GetitemOp` and `ReshapeOp`. How about something like this:
```py
class TriangularSolveMeta(type):
    ...implement caching as in GetitemMeta...

class TriangularSolveOp(Op, metaclass=TriangularSolveMeta):
    def __init__(self, upper, transpose):
        assert isinstance(upper, bool)
        assert isinstance(transpose, bool)
        self.upper = upper
        self.transpose = transpose
    def __call__(self, x, y):
        raise NotImplementedError
```
then later you can, in torch.py
```py
for upper in [False, True]:
    for transpose in [False, True]:
        @TriangularSolveOp(upper, transpose).register(Tensor, Tensor)
        def triangular_solve(x, y):
            return  x.data.triangular_solve(y.data, upper, transpose)
```

Reply:
My intention is to use this op on `torch.Tensor` or `numpy.ndarray`, so it might be better to keep the current pattern as-is.

Reply:
Ok, but we'll want to change to `TriangularSoveOp` eventually...